firmware/quadruna/BMS/src/app/app_soc.c

```diff
@@ -180,3 +202,66 @@ void app_soc_resetSocCustomValue(float soc_percent)
     stats.is_corrupt = true;
     app_canTx_BMS_SocCorrupt_set(stats.is_corrupt);
 }
+
+bool app_soc_readSocFromSd(float *saved_soc_c)
+{
+    uint8_t sd_read_data[HW_DEVICE_SECTOR_SIZE];
+    uint8_t sd_read_soc_bytes[NUM_SOC_BYTES];
+    uint8_t sd_read_crc_bytes[NUM_SOC_CRC_BYTES];
+
+    uint32_t sd_read_soc;
+    uint32_t sd_read_crc;
+    float soc;
+
+    if(hw_sd_read(sd_read_data, DEFAULT_SOC_ADDR, 1) == SD_CARD_OK)
+    {
+        memcpy(sd_read_soc_bytes, sd_read_data, sizeof(uint32_t));
+        sd_read_soc = ARRAY_TO_UINT32(sd_read_soc_bytes);
+        soc = convert_bytes_to_float(sd_read_soc_bytes);
+    }
+    else
+    {
+        *saved_soc_c = -1.0f;
+        return false;
+    }
+
+    //CRC bytes are stored right after the SOC bytes
+    if(hw_sd_read(sd_read_data, DEFAULT_SOC_ADDR + NUM_SOC_BYTES, 1) == SD_CARD_OK)
+    {
+        memcpy(sd_read_crc_bytes, sd_read_data, sizeof(uint32_t));
+        sd_read_crc = ARRAY_TO_UINT32(sd_read_crc_bytes);
+    }
+    else
+    {
+        *saved_soc_c = -1.0f;
+        return false;
+    }
+
+    uint32_t calculated_crc = hw_crc_calculate(&sd_read_soc, 1);
+
+    if(calculated_crc == sd_read_crc)
+    {
+        *saved_soc_c = soc;
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+void app_soc_writeSocToSd(float soc)
+{
+    uint8_t sd_write_data[HW_DEVICE_SECTOR_SIZE];
+    convert_float_to_bytes(sd_write_data, soc);
+    uint32_t soc_value = ARRAY_TO_UINT32(sd_write_data);
+
+#ifdef TARGET_EMBEDDED
```

Comment:
Why not wrap the entire function? Is there a testable benefit to leaving the above 3 lines in?